File: DART/binary_search.dart
void main() {
  List<int> arr = [0, 2, 3, 5, 7, 11, 13, 17];
  int userVal = 3;
  int min = 0;
  int max = arr.length - 1;
  bst(arr, userVal, min, max);
}

//binary search tree
bst(List<int> arr, int userVal, int min, int max) {
  if (max >= min) {
    print('min $min');
    print('max $max');
    int mid = ((max + min) / 2).floor();
    if (userVal == arr[mid]) {
      print('Your value is at index: ${mid}');
    } else if (userVal > arr[mid]) {
      bst(arr, userVal, mid + 1, max);
    } else {
      bst(arr, userVal, min, mid - 1);
    }
  }
  return null;
}

File: C/Binary_Search.c
// C program to implement iterative Binary Search 
#include <stdio.h> 
  
// A iterative binary search function. It returns 
// location of x in given array arr[l..r] if present, 
// otherwise -1 
int binarySearch(int arr[], int l, int r, int x) 
{ 
    while (l <= r) { 
        int m = l + (r) / 2; 
  
        // Check if x is present at mid 
        if (arr[m] == x) 
            return m; 
  
        // If x greater, ignore left half 
        if (arr[m] < x) 
            l = m + 1; 
  
        // If x is smaller, ignore right half 
        else
            r = m - 1; 
    } 
  
    // if we reach here, then element was 
    // not present 
    return -1; 
} 
  
int main(void) 
{ 
    printf("Enter size of the array you want to search from : \n"); 
    int n;
    scanf("%d",&n);
    printf("Enter the array : \n");
    int arr[n];
    for(int i = 0 ; i < n ; i++)
    {
      scanf("%d",&arr[i]);
    }
    int x;
    printf("Enter the element you want to search from in the array \n");
    scanf("%d",&x);
  //Calling the binarySearch function
    int result = binarySearch(arr, 0, n-1, x); 
    (result == -1) ? printf("Element is not present"
                            " in array") 
                   : printf("Element is present at "
                            "index %d", 
                            result); 
    return 0; 
} 
